### DISCUSSION TITLE
Is it possible to personalize the observation function in python?

### DISCUSSION THREAD

**Discussion by Wenbo11** [ds4dm/ecole#87] [Questions] [answered]
Hey,

The GCNN branching tutorial samples observation by the build-in functions: ecole.observation.StrongBranchingScores() and ecole.observation.Pseudocosts(). Is it possible to personalize the observation function (branching policy) for ecole in python? Thank you so much!

Best,
Wenbo

**Comment by AntoinePrv** [comment, marked as answer]
Hi,

The observation function is an object that extract data from SCIP, typically to get the data you would feed to the ML model.
It  can be done in Python or in C++, either by inheriting of an existing one (if you need some of its data), or from scratch. You can see more details here: https://doc.ecole.ai/master/howto/create-functions.html 

Ecole implements a few observation functions that we deemed valuable to many users (hopefully more to come).
In the case of `Pseudocosts` and `StrongBranchingScores`, they are observation functions.
As such, they do not perform any branching, only extract data, even if they are related to a branching policy.
They are useful to get "expert" data and reproduce such branching rules (by taking the argmax of the scores and passing it to `env.step`).

Hope this answer your question. Let us know if we can elaborate on anything.

**Comment by Wenbo11** [comment]
Thanks for your response! It helps a lot. I have a small suggestion for the documents of [**dynamics_kwargs](https://doc.ecole.ai/master/reference/environments.html#ecole.environment.Branching). I guess it could be helpful to detail it in the document. [For instance](https://github.com/ds4dm/ecole/blob/efd60587b26fc08da68754395d20974c29b0fd15/libecole/src/environment/branching-dynamics.cpp#L17), if I want to query the pseudobranching candidates rather than the LPbranching candidates, I can directly pass pseudo_candidates=True when initializing the environment.

**Comment by AntoinePrv** [comment]
Thanks for the feedback.
We documented this in the [Environment parameters](https://doc.ecole.ai/master/using-environments.html#environment-parameters) section

> Each environment can be given a set of parameters at construction, in order to further customize the task being solved. For instance, the Branching environment takes a `pseudo_candidates` boolean parameter, to decide whether branching candidates should include all non fixed integral variables, or only the fractional ones.

In your opinion, is there a way we can improve on this?
Perhaps we could document in the [Creating environment](https://doc.ecole.ai/master/howto/create-environments.html) section that `Dynamics` constructor argument are forwarded from `Environmnents` extra arguments

**Comment by Wenbo11** [comment]
Yes, I personally believe it can be helpful to document it in the Creating environment section.

**Comment by AntoinePrv** [comment]
I have added [another section](https://doc.ecole.ai/v0.2.x/howto/create-environments.html#passing-parameters) to the doc
